pandas/core/generic.py

```diff
@@ -2774,40 +2774,28 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
         # set the default here, so functions examining the signaure
         # can detect if something was set (e.g. in groupby) (GH9221)
         if axis is None:
-            axis = 0
+            axis = self._stat_axis_name
         axis = self._get_axis_number(axis)
-        method = mis._clean_fill_method(method)
+        method = missing._clean_fill_method(method)
 
         from pandas import DataFrame
         if value is None:
             if method is None:
                 raise ValueError('must specify a fill method or value')
-            if self._is_mixed_type and axis == 1:
+            if self._is_mixed_type:
+                if (self.ndim > 2) and (axis == 0):
+                    raise NotImplementedError('cannot fill across axis 0 for mixed dtypes')
```

Comment:
The way filling across mixed-dtype blocks for 2D frames is done is by transposing axes. This is actually what I proposed doing in my original PR from last year to extend this filling across higher dimensionality frames, but you rejected that idea. I could fill 2D frames using the same general method I'm proposing here, but that will result in in some backwards incompatible behavior and users that depending on being able to fill across mixed-type blocks for 2D would  no longer be able to do so.